File: miscellaneous/sorting/merge-sort/20230330/merge_sort.py
from typing import List
from math import floor


def sort(nums: List[int]) -> List[int]:
    length = len(nums)

    if length < 2:
        return nums

    mid = floor(length / 2)
    return merge(sort(nums[0:mid]), sort(nums[mid:length]));


def merge(left, right):
    merged = []

    while left and right:
        merged.append(left.pop(0) if left[0] < right[0] else right.pop(0))

    return merged + left + right

File: miscellaneous/sorting/merge-sort/20230330/merge-sort.php
<?php

declare(strict_types=1);

function mergeSort(array $nums): array
{
    if (count($nums) <= 1) return $nums;
    $mid = (int) floor(count($nums) / 2);
    return merge(
        mergeSort(array_slice($nums, 0, $mid)),
        mergeSort(array_slice($nums, $mid))
    );
}

function merge(array $left, array $right): array
{
    $merged = [];

    while (count($left) && count($right)) {
        $merged[] = $left[0] < $right[0] ? array_shift($left) : array_shift($right);
    }

    return [...$merged, ...$left, ...$right];
}